### PR TITLE
[android] make promiseimpl's reject method thread safe

### PR DESCRIPTION
**Pull request by shangtee**
## Summary

`PromiseImpl` is not thread-safe, causing crashes on accessing a variable that can become null on a different thread. This has been causing Discord's `BillingManagerModule` to throw `NullPointerException`s about 800-1k times each day (estimated with sentry sampling). 

Also see:
https://github.com/facebook/react-native/issues/35068 
https://app.asana.com/0/1202593986197760/1203652454773960
https://discord.sentry.io/issues/3504368273/?project=5992375

## Changelog

[Android] [Fixed] - Fixed PromiseImpl's reject not thread-safe

### CODE REVIEW

**Comment by dhleong** [comment]
Seems reasonable to me; should we not also synchronize `resolve`?

**Comment by dhleong** [comment]
I should also ask—would it make sense to find the source of the race and address there? This will likely solve the crash, but the race might result in other problems...

**Comment by dhleong** [comment]
Looking at the source, I'm actually even more confused why there would be a race here. I don't imagine the Promise is somehow shared upstream, and it doesn't appear that there are any paths in there to spawn multiple threads that would have contention for it. Can you repro the race? Something strange seems to be going on 🤔

**Comment by shangtee** [comment]
@dhleong I think it makes sense to find the source of the race. I'm not too familiar with the react code for billing or android in general; do you think you or someone else can pair debug with me? 🙇

**Comment by dhleong** [comment]
@shangtee Let's continue the discussion in `#android`! Miguel and I were batting some ideas back and forth starting here: https://canary.discord.com/channels/281683040739262465/270710046239752193/1083452524079611946.

tl;dr is we suspect the billing library is doing something sus; our billing manager code doesn't look like it's doing anything that would cause contention. The sentry stack looks like it's triggering from the old, deprecated SKU fetch flow, so it's possibly old library shenanigans
